### PR TITLE
Added ArrayStack.

### PR DESCRIPTION
**Pull request by cseed**
Use in RegionValueBuilder.

### CODE REVIEW

**Comment by danking** [comment]
This quote from the 2.12.0 collections docs is as depressing and still seemingly misguided:

> (Since version 2.12.0) mutable.Stack is an inelegant and potentially poorly-performing wrapper around List. Use a List assigned to a var instead.

File: src/main/scala/is/hail/utils/ArrayStack.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    a(size_) = x
+    size_ += 1
+  }
+
+  def pop(): T = {
+    assert(size_ > 0)
+    size_ -= 1
+    val x = a(size_)
+    a(size_) = uninitialized[T]
```
Comment (danking) [inline src/main/scala/is/hail/utils/ArrayStack.scala]: I think this is unnecessary. The privacy of a and the assert on `apply` makes the value at this position effective unknowable.

File: src/test/scala/is/hail/utils/ArrayStackSuite.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    assert(s.top == -13)
+
+    assert(s(0) == -13)
+    assert(s(1) == 11)
+    assert(s(2) == -1)
+    assert(s(3) == 0)
+    assert(s(4) == 13)
+
+    s(2) = 39
+    assert(s.pop() == -13)
```
Comment (danking) [inline src/test/scala/is/hail/utils/ArrayStackSuite.scala]: We should add a test of `top` after a call to `pop`.

File: src/main/scala/is/hail/utils/ArrayStack.scala (hunk 1 above)
Comment (cseed) [reply src/main/scala/is/hail/utils/ArrayStack.scala]: This is needed to let popped stack elements get garbage collected.

Comment (tpoterba) [reply src/main/scala/is/hail/utils/ArrayStack.scala]: won't this box?

Comment (cseed) [reply src/main/scala/is/hail/utils/ArrayStack.scala]: I made uninitialized (which was still allocating?!) specialized.

**Comment by cseed** [comment]
Back to you.

**Pull request merged**